internal/proxy_service/tcp/proxy_service_test.go

```diff
@@ -0,0 +1,348 @@
+
+	"github.com/cyberark/secretless-broker/internal/proxy_service/mock"
+)
+
+const testString1 = "good heavens"
+const testString2 = "moderate heavens"
+const testString3 = "best of heavens"
+
+func TestNewProxyService(t *testing.T) {
+	t.Run("empty constructor arguments result in errors", func(t *testing.T) {
+		_, err := NewProxyService(
+			nil,
+			nil,
+			nil,
+			nil,
+		)
+		assert.Error(t, err)
+		if err == nil {
+			return
+		}
+	})
+
+	t.Run("non-empty constructor arguments result in no error", func(t *testing.T) {
+		_, err := NewProxyService(
+			mock.NewConnector().Connect,
+			mock.NewListener(),
+			mock.NewLogger(),
+			mock.NewCredentialRetriever().RetrieveCredentials,
+		)
+		assert.NoError(t, err)
+		if err != nil {
+			return
+		}
+	})
+
+	t.Run("zeroizes credentials from retrieveCredentials", func(t *testing.T) {
+		// prepare
+		connector := mock.NewConnector()
+
+		credentialRetriever := mock.NewCredentialRetriever()
+		creds := map[string][]byte{
+			"credName": []byte("credValue"),
+		}
+		credentialRetriever.On("RetrieveCredentials").Return(
+			creds,
+			fmt.Errorf("couldn't retrieve credentials"))
+
+		listener := mock.NewListener()
+
+		// exercise
+		ps := proxyService{
+			connector:           connector.Connect,
+			retrieveCredentials: credentialRetriever.RetrieveCredentials,
+			listener:            listener,
+		}
+		err := ps.handleConnection(nil)
+
+		// assert
+		assert.Error(t, err)
+		if err == nil {
+			return
+		}
+		assert.Equal(
+			t,
+			string(creds["credName"]),
+			strings.Repeat("\x00", len([]byte("credValue"))),
+		)
+	})
+}
+
+func TestProxyService_Start(t *testing.T) {
+	t.Run("stopped proxy service cannot be restarted", func(t *testing.T) {
+		connector := mock.NewConnector()
+		credentialRetriever := mock.NewCredentialRetriever()
+		listener := mock.NewListener()
+		listener.On("Close").Return(nil)
+		logger := mock.NewLogger()
+
+		ps, _ := NewProxyService(
+			connector.Connect,
+			listener,
+			logger,
+			credentialRetriever.RetrieveCredentials,
+		)
+
+		err := ps.Stop()
+		assert.NoError(t, err)
+		if err != nil {
+			return
+		}
+
+		err = ps.Start()
+		assert.Error(t, err)
+	})
+
+	t.Run("propagates error from Accept", func(t *testing.T) {
+		// prepare
+		logger := mock.NewLogger()
+		logger.On("Errorf").Return()
+
+		connector := mock.NewConnector()
+		connector.On("Connect").Return(nil, nil)
+
+		credentialRetriever := mock.NewCredentialRetriever()
+		credentialRetriever.On(
+			"RetrieveCredentials",
+		).Return(nil, nil)
+
+		listener := mock.NewListener()
+		listener.On("Accept").Return(nil, fmt.Errorf("some error"))
+
+		// exercise
+
+		ps, err := NewProxyService(
+			connector.Connect,
+			listener,
+			logger,
+			credentialRetriever.RetrieveCredentials)
+		err = ps.Start()
+
+		// sanity assert
+		assert.NoError(t, err)
+		if err != nil {
+			return
+		}
+
+		// artificial sleep needed to wait for Errorf
+		time.Sleep(time.Millisecond)
```

Comment:
We don't need to do this now, but it's worth pointing out: One of the great things bout mocks is we totally control them.  So you could easily add a channel to the logger mock (this would change only the interface of the mock ctor, not the thing being mocked), and then use that to synchronize here without the indeterminacy of sleep.

Reply:
Great, I was wrecking my brain about this one. Could you provide an example please, that'd help a whole bunch...